src/computer_player.c

```diff
@@ -0,0 +1,604 @@
+// computer algrithm for hard level computer
+//
+// features:
+// - Count cards on the table, record all the played card                                               (x)
+// - Quickly get rid of matched numbers in the hand.                                                    (x)
+// - Keeping at most one Wild Card until the very end.                                                  (x)
+// - use reverse card when you have no card to play                                                     (x)
+// - Play series of similar color cards first, use draw 2 card,                                         (x)
+//   skip card, or reverse card first based on the most color that computer has                         (x)
+// - When playing a Wild Card +4, always change the color based on what the computer has the most       (x)
+// - Try to keep at least one +2 card in your hand.         (x)                                         (x)
+// - Play higher number cards first to get more points      (x)                                         (x)
+// - Note the last card color when players have one card left and avoid to play it.
+//
+// input :
+// - card on the table (for update and play)
+// - if player "UNO"
+// - card drawed from the deck
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "cards_management.h"
+ 
+#define MAX 40
+#define hand_MAX 20
+#define plarCardNumber 10
+
+///* (not used for now) */
+//struct CARD record_array[MAX]; // array for recording all the cards
+//int record_size = 0, handcard_size = 0;
+//
+//void record(struct CARD card)
+//{
+//    record_size++;
+//    record_array[record_size - 1] = card;
+//}
+
+/* 
+* @brief: function that find the most occurrence with the color 
+* 
+* @return CardColor the color with moset occurrence
+*/
+enum CardColor  find_most_color(struct DECK *cards_on_hand)
+{
+    int R_count = 0, G_count = 0, B_count = 0, Y_count = 0, most = 0, least = 0;
+    struct DECK *temp;
+    struct DECK *head = cards_on_hand;
+    if (head == NULL)
+    {
+        printf("hand card is empty!");
+        return 0;
+    }
+    temp = head;
+    while (temp != NULL)
+    {
+        if (temp->card.color == RED)
+        {
+            R_count++;
+        }
+        else if (temp->card.color == BLUE)
+        {
+            B_count++;
+        }
+        else if (temp->card.color == GREEN)
+        {
+            G_count++;
+        }
+        else
+        {
+            Y_count++;
+        }
+        temp = temp->next;
+    }
+    if (R_count > B_count && R_count > G_count && R_count > Y_count)
+    {
+        return RED;
+    }
+    else if (B_count > R_count && B_count > G_count && B_count > Y_count)
+    {
+        return BLUE;
+    }
+    else if (G_count > R_count && G_count > B_count && G_count > Y_count)
+    {
+        return GREEN;
+    }
+    else if (Y_count > R_count && Y_count > B_count && Y_count > G_count)
+    {
+        return YELLOW;
+    }
```

Comment:
It's better to have a variadic function int get_max (n, ...) to simply the if(), else if () here.
int max = max_value (4,R_count ,G_count ,B_count ,Y_count );

int max_value (int n,...) {
    int max = INT_MIN;
    int temp;  
    va_list list;
    va_start(list,n);

    for(int i = 0;i < n; i++){
        temp = va_arg(list,int);
        if(max < temp){
            max = temp;
        }  
    }
    va_end(list);
    return max;
}